### DISCUSSION TITLE
Python API - Numpy dependency

### DISCUSSION THREAD

**Discussion by fedehann** [duckdb/duckdb#5478] [Ideas]
Hi, I was trying to compile a **client-side GUI-application** developed in Python (packaging the latter relying on [Nuitka](https://github.com/Nuitka/Nuitka)). The application uses DuckDb as a computational engine for some basic financial calculations. This involves launching 20 SQL queries on downloaded datasets which are quite large (tens of millions of records), retrieving params from the GUI. I noticed, however, that the **Python API depends on Numpy**, which makes the compiled application quite big for its scope (about 130Mbs, half of which is Numpy-stuff). 
Is the Numpy dependency mandatory for simply managing connections, cursors, running SQL statements and fetching relatively modest amount of data through the Python API? **Is there a way to make this dependency optional** (for these basic requirements only)?

**Comment by hannes** [comment]
Its indeed not always required, happy to review a PR that makes it optional

**Comment by slonik-az** [comment]
> This involves launching 20 SQL queries on downloaded datasets which are quite large (tens of millions of records), 

I am curious to understand how freeing meager 130MB consumed by `numpy` would solve your memory problem when you are dealing with datasets of that size.

**Reply by fedehann** [reply]
Well, it is more a storage thing than a memory one: if I ship an application, in the package I don't provide the data (which amounts to 5-6 GBs). That is handled by the user, who uploads such datasets; the application then processes the data, runs some aggregations and produces an output. So, the shipped package is unnecessarily large (because of Numpy). It is not a big issue, but it would be nice to avoid dependencies which are not needed in basic scenarios. Something like SQLite (which I extensively use) only has a basic "driver", which is enough to communicate with the underlying library: that is included in Python's stdlib and has no other dependencies. I understand that DuckDb is primarily intended for data-science workflows requiring the interaction with pandas, numpy, scikit-learn, etc, but it would be useful in my situation as well, where Numpy's overhead is not needed at all.

**Comment by Mause** [comment]
Can you explain what you're measuring as "numpy stuff"? The numpy wheel itself is only around 14mb, and even expanded it's only 60mb, nowhere near the 130mb you quoted

**Reply by fedehann** [reply]
Hi Mause, yeah, the expanded Numpy-wheel is about 60mb, largely due to openblas:

> I noticed, however, that the Python API depends on Numpy, which makes the compiled application quite big for its scope (about 130Mbs, half of which is Numpy-stuff)

**Comment by slonik-az** [comment]
Well, fighting software bloat is, of course, a noble goal. But numpy is quite a foundational part of python data-science toolbox. Removing numpy will invalidate pandas, scikit-learn and many other python libraries. The only realistic case of having bare-bones DuckDB is within the embedded space where every byte of RAM or disk are at premium. Other than that making numpy optional, IMHO, would be lot of work for little gain.

EDIT: By lot of work I mean not only changes to the code base itself but also changes to the tests that now have to separately test both cases with and without numpy.

**Reply by fedehann** [reply]
I understand and definitely agree on Numpy being essential for anything data-science related (in the Python world), but it is often required, for instance in corporate applications, to implement quick and dirty elaboration processes as minimal as possible in scope and size, to put together data from different providers, run aggregations, joins and other trivial DB-operations, meeting very specific reporting needs - or for testing. DuckDb would be perfect for this kind of use-cases (and would become a real and better alternative to SQLite, at least to me) especially without the Numpy dependency. Once again: it's a nice-to-have more than a strict requirement: the current Python API just does too much for my needs, but I recognize that this fits a potentially small niche - even if not as small as you may think.

To be clear: I am not proposing at all to remove the support for pandas, pyarrow, scikit, etc, but rather to make the dependency on Numpy optional for the trivial things (pure interaction with the DBMS, raw, unoptimized data retrieval). If you take for instance the official Python driver for SQLite, it's a really simple thing, included in the standard library, and with no dependencies at all.

**Reply by FrancoisLepoutre** [reply]
Hi Fedehann,

Thank you - and others - for this post and the related thread:-)

I understand your wish to have a nicely embeddable python+duckdb based application. Same case here! Not as advanced as you are, though and I'll read carefully what you said in this thread and possibly in coming ones as well.

We ran a short POC python+qt+duckdb+pyarrow+numpy+numba encapsulation on windows 10-11 here via pyinstaller. After a few tests - and bloatware dropping within pyinstaller! - the full executable was 110 Mb (and that includes the numba-supporting virtual machine and additional libraries). I cannot exactly assess what size numpy is within the package. But I wonder whether it is as big as what you got within nuitka.

We used only very basic numpy functionality, as an export output format for duckdb. We then used and abused numba which was absolutely splendid to work with. And then we imported the massaged output back to content into great duckdb, fast as well. 

Both duckdb and numba proved brilliant at this.  I am comfy with sql but certainly no numpy expert. But as far as I could judge the numpy dependency proved light enough in our view. I guess we had only a partial numpy (no openblas?). I'll investigate when I can find the time to check this POC stuff. But I do not feel that the numpy dependency is so big to be issue. As you said Pandas would certainly be a big liability. But I wonder if the problem could not lie possibly also partly within nuitka?  and I may be wrong on that for sure!

I was really glad to hear you are using nuitka. And I really look forward to your feedaback on the python+qt+nuitka combo as I share your willingness to build a compact python+qt  building environment for end-user desktop big-data-centric application!

**Reply by fedehann** [reply]
Hi Francois,
I'm very happy to hear you're doing something similar to us! Compact, embedded applications are very useful indeed, and to have the chance of quickly developing one with a high-level language such as Python and with new powerful tools (such as DuckDb) is really really cool.

**Off topic:**

I am absolutely no expert on the topic, so take my words with a grain of salt, but Nuitka and Pyinstaller are quite different things; Nuitka is a fully-fledged compiler, so it basically takes Python code, turns it into C, and then calls a C compiler to do its job (GCC, etc). Pyinstaller, on the other end, takes the Python interpreter, takes the related Python code, and puts all together in a bundle.

So, when you run an application bundled with Pyinstaller, you are basically running Python code on top of the bundled interpreter; when you run an application compiled with Nuitka, you are running machine code directly (still, with all the overhead required by migrating Python to C). The advantage of Nuitka-applications is that they run slightly faster and they are much less susceptible to antivirus, etc (an AV doesn't like if you bootstrap an interpreter and its code together in a binary). The disadvantage is that they are bigger, since the code is mostly converted from Python to machine-code. This is why Pyinstaller produces much more compact executables than Nuitka.

**Reply by FrancoisLepoutre** [reply]
Thanks for the feedback!

**Just Python-Related Even More Off topic:**

_[...the code is mostly converted from Python to machine-code. This is why Pyinstaller produces much more compact executables than Nuitka.]_

I clearly understood the difference between both solutions and am following the development of Nuitka with great attention. But I missed the fact that the nuitka compilation test would trigger such a significant expansion in executable size. Thanks for this useful info.

Yep i agree AV software are specifically a nuisance when working with bundled packages à la pyinstaller. Already had serious AV issues on another legacy interpreter with another packager  even when they are properly signed! So I appreciate the impressive work done by Kay Hayen and others. We need python compilers for desktop deployment. 

At the current moment however, we have a specific need for numpy data massaging (when our great duck sql is not really applicable). And I just wondered how we could have our beginners' numpy code faster within our pyinstaller POC. The answer was plain easy: with an embedded numba and a few "decorator-based" tricks. 

The addition of this numpy+numba combo (on top of pythonqt-duckdb-pyarrow...) proved incredibly effective at getting specific python-numpy-based code fast. As python dummies we were able reach to-the-metal llvm-based speed levels including parallelization. No julia, no C...  

Of course since numba is clearly incompatible with a "just-in-time compiler", if we need to bring nuitka into the loop, we will need to move the set of our optimizations tricks onto some sort of "ahead of time compiler" such pythran. And then push the generated so/dll output into nuitka.  

As “premature optimization is the root of all evil”, I'll wait until we have some feedback on this "nuitka+pythran" combination to try this great nuitka+QT+duckdb combination! 

My dream for duckdb in the very long run: some sort of llvm-based engine stored procs language support crunching operations  à la numba on the engine vector-based content. That's of course just a vaporware dream as we have already so much at hand to-day!

Thanks to this great duckdb project.
